core/shared/platform/windows/win_file.c

```diff
@@ -4,117 +4,766 @@
+                && (win_handle)->raw.handle == INVALID_HANDLE_VALUE)) { \
+            return __WASI_EBADF;                                        \
+        }                                                               \
+    } while (0)
+
+#define CHECK_VALID_FILE_HANDLE(win_handle)                          \
+    do {                                                             \
+        if ((win_handle) == NULL) {                                  \
+            return __WASI_EBADF;                                     \
+        }                                                            \
+        if ((win_handle)->type == windows_handle_type_socket) {      \
+            return __WASI_EINVAL;                                    \
+        }                                                            \
+        if (((win_handle)->type == windows_handle_type_file          \
+             && (win_handle)->raw.handle == INVALID_HANDLE_VALUE)) { \
+            return __WASI_EBADF;                                     \
+        }                                                            \
+    } while (0)
+
+#define CHECK_VALID_WIN_DIR_STREAM(win_dir_stream)         \
+    do {                                                   \
+        if ((win_dir_stream) == NULL)                      \
+            return __WASI_EINVAL;                          \
+        CHECK_VALID_FILE_HANDLE((win_dir_stream)->handle); \
+    } while (0)
+
+#define CHECK_VALID_PATH(path)          \
+    do {                                \
+        if (strstr(path, "\\") != NULL) \
+            return __WASI_EINVAL;       \
+    } while (0)
+
+static __wasi_errno_t
+convert_winsock_error_code(int error_code)
+{
+    switch (error_code) {
+        case WSASYSNOTREADY:
+        case WSAEWOULDBLOCK:
+            return __WASI_EAGAIN;
+        case WSAVERNOTSUPPORTED:
+            return __WASI_ENOTSUP;
+        case WSAEINPROGRESS:
+            return __WASI_EINPROGRESS;
+        case WSAEPROCLIM:
+            return __WASI_EBUSY;
+        case WSAEFAULT:
+            return __WASI_EFAULT;
+        case WSAENETDOWN:
+            return __WASI_ENETDOWN;
+        case WSAENOTSOCK:
+            return __WASI_ENOTSOCK;
+        case WSAEINTR:
+            return __WASI_EINTR;
+        case WSAEAFNOSUPPORT:
+            return __WASI_EAFNOSUPPORT;
+        case WSAEMFILE:
+            return __WASI_ENFILE;
+        case WSAEINVAL:
+            return __WASI_EINVAL;
+        case WSAENOBUFS:
+            return __WASI_ENOBUFS;
+        case WSAEPROTONOSUPPORT:
+            return __WASI_EPROTONOSUPPORT;
+        case WSAEPROTOTYPE:
+            return __WASI_EPROTOTYPE;
+        case WSAESOCKTNOSUPPORT:
+            return __WASI_ENOTSUP;
+        case WSAEINVALIDPROCTABLE:
+        case WSAEINVALIDPROVIDER:
+        case WSAEPROVIDERFAILEDINIT:
+        case WSANOTINITIALISED:
+        default:
+            return __WASI_ENOSYS;
```

Comment:
I don't think `ENOSYS` is the correct one here, that suggests the function is not implemented, whereas it is, it just returns an error. Maybe `EINVAL` would be better instead?